{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phylogenetic analysis\n",
    "To investigate further the functional roles and evolutionary origin of your candidate (upregulated) genes, you decide to perform a phylogenetic analysis comparing them against other prokaryotic proteomes (including the reference genome of the species matching our isolated strain in the hot spring).‚Äù\n",
    "\n",
    "### Tasks:\n",
    "For each over expressed gene (you have individual FASTA files created from previous section), run a standard phylogenetic workflow: \n",
    "1. Run a blast search for each over expressed protein against all reference proteomes\n",
    "2. Extract hits with e-value <= 0.00001 (tip: you can use blast parameters for this)\n",
    "3. Create a FASTA file with all the sequences of selected hits  (tip: you can use the extract_sequences_from_blast_result.py)\n",
    "4. Build a phylogenetic tree out of the fastA file (suggested tools: clustalo, iqtree)\n",
    "5. Visualize the result (suggested tools: etetoolkit.org/treeview, itol.embl.de, ete3)\n",
    "6. Using the additional info file located at  additional_seq_info.tsv (see full path above) extract functional and taxonomic information for each homolog in the trees, and visualize it in the tree for better interpretation.  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. BLAST\n",
    "The first thing to do is building the BLAST database using all reference proteomes from the file `/home/2019_2020/data/phylo/all_ref_proteomes.faa`  <br><br>\n",
    "We will have to extract the four sequences of the over-expressed genes from `/home/2019_2020/data/phylo/novel_proteome.faa` <br><br>\n",
    "Then we can do the BLAST and extract the best hits with the python script provided as `python extract_seqs_from_blast_result.py  blast_output all_ref_proteomes.faa > homologs.faa`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NP_213724.1.fas  NP_Unk02.fas               all_ref_proteomes.faa.pin\n",
      "NP_213887.1.fas  all_ref_proteomes.faa      all_ref_proteomes.faa.psq\n",
      "NP_Unk01.fas     all_ref_proteomes.faa.phr  \u001b[0m\u001b[38;5;34mextract_seqs_from_blast_result.py\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "cd \n",
    "#mkdir ./phylo\n",
    "cd ./phylo\n",
    "#cp /home/2019_2020/data/phylo/extract_seqs_from_blast_result.py .\n",
    "#cp /home/2019_2020/data/phylo/all_ref_proteomes.faa .\n",
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 01/06/2020 22:00:04\n",
      "New DB name:   /home/2019_2020/s.sanchez-heredero/phylo/all_ref_proteomes.faa\n",
      "New DB title:  all_ref_proteomes.faa\n",
      "Sequence type: Protein\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 88473 sequences in 5.04095 seconds.\n"
     ]
    }
   ],
   "source": [
    "# Make database\n",
    "makeblastdb -dbtype prot -in all_ref_proteomes.faa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract my sequences\n",
    "cat /home/2019_2020/data/phylo/novel_proteome.faa | grep -A 1 'NP_213724.1' > NP_213724.1.fas\n",
    "cat /home/2019_2020/data/phylo/novel_proteome.faa | grep -A 1 'NP_213887.1' > NP_213887.1.fas\n",
    "cat /home/2019_2020/data/phylo/novel_proteome.faa | grep -A 1 'NP_Unk01' > NP_Unk01.fas\n",
    "cat /home/2019_2020/data/phylo/novel_proteome.faa | grep -A 1 'NP_Unk02' > NP_Unk02.fas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BLAST\n",
    "blastp -task blastp -query NP_213724.1.fas -db all_ref_proteomes.faa -outfmt 6 -evalue 0.00001 > NP_213724.1.blast\n",
    "blastp -task blastp -query NP_213887.1.fas -db all_ref_proteomes.faa -outfmt 6 -evalue 0.00001 > NP_213887.1.blast\n",
    "blastp -task blastp -query NP_Unk01.fas -db all_ref_proteomes.faa -outfmt 6 -evalue 0.00001 > NP_Unk01.blast\n",
    "blastp -task blastp -query NP_Unk02.fas -db all_ref_proteomes.faa -outfmt 6 -evalue 0.00001 > NP_Unk02.blast\n",
    "\n",
    "python extract_seqs_from_blast_result.py NP_213724.1.blast all_ref_proteomes.faa > NP_213724.1.blast_homologs.faa\n",
    "python extract_seqs_from_blast_result.py NP_213887.1.blast all_ref_proteomes.faa > NP_213887.1.blast_homologs.faa\n",
    "python extract_seqs_from_blast_result.py NP_Unk01.blast all_ref_proteomes.faa > NP_Unk01.blast_homologs.faa\n",
    "python extract_seqs_from_blast_result.py NP_Unk02.blast all_ref_proteomes.faa > NP_Unk02.blast_homologs.faa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NP_213724.1.blast               NP_Unk02.blast\n",
      "NP_213724.1.blast_homologs.faa  NP_Unk02.blast_homologs.faa\n",
      "NP_213724.1.fas                 NP_Unk02.fas\n",
      "NP_213887.1.blast               all_ref_proteomes.faa\n",
      "NP_213887.1.blast_homologs.faa  all_ref_proteomes.faa.phr\n",
      "NP_213887.1.fas                 all_ref_proteomes.faa.pin\n",
      "NP_Unk01.blast                  all_ref_proteomes.faa.psq\n",
      "NP_Unk01.blast_homologs.faa     \u001b[0m\u001b[38;5;34mextract_seqs_from_blast_result.py\u001b[0m\n",
      "NP_Unk01.fas\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br><br>\n",
    "### 2. Building the trees\n",
    "Now that we have the sequences that we want to use for the phylogenetic analysis, we can do the phylogenetic analysis, for which we will need to:\n",
    "1. Make an alignment of the sequences using `clustalo`\n",
    "2. Build the tree using `iqtree`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make alignmets\n",
    "/home/miniconda3/bin/clustalo -i NP_213724.1.blast_homologs.faa > NP_213724.1.blast_homologs.alg\n",
    "/home/miniconda3/bin/clustalo -i NP_213887.1.blast_homologs.faa > NP_213887.1.blast_homologs.alg\n",
    "/home/miniconda3/bin/clustalo -i NP_Unk01.blast_homologs.faa > NP_Unk01.blast_homologs.alg\n",
    "/home/miniconda3/bin/clustalo -i NP_Unk02.blast_homologs.faa > NP_Unk02.blast_homologs.alg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 1.6.12 for Linux 64-bit built Aug 15 2019\n",
      "Developed by Bui Quang Minh, Nguyen Lam Tung, Olga Chernomor,\n",
      "Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    localhost.localdomain (SSE3, 125 GB RAM)\n",
      "Command: /home/miniconda3/bin/iqtree -s NP_213724.1.blast_homologs.alg\n",
      "Seed:    323297 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Mon Jan  6 23:03:24 2020\n",
      "Kernel:  SSE2 - 1 threads (40 CPU cores detected)\n",
      "\n",
      "HINT: Use -nt option to specify number of threads because your CPU has 40 cores!\n",
      "HINT: -nt AUTO will automatically determine the best number of threads to use.\n",
      "\n",
      "Reading alignment file NP_213724.1.blast_homologs.alg ... Fasta format detected\n",
      "Alignment most likely contains protein sequences\n",
      "Alignment has 36 sequences with 563 columns, 466 distinct patterns\n",
      "300 parsimony-informative, 107 singleton sites, 156 constant sites\n",
      "                     Gap/Ambiguity  Composition  p-value\n",
      "   1  122586.Q9JX95_NEIMB   40.67%    passed     96.59%\n",
      "   2  122586.Q9K1E8_NEIMB   39.08%    passed     98.58%\n",
      "   3  85962.O25902_HELPY    41.39%    failed      2.99%\n",
      "   4  85962.G3P_HELPY       41.03%    failed      2.58%\n",
      "   5  272561.G3P_CHLTR      40.67%    passed     95.22%\n",
      "   6  251221.Q7NG48_GLOVI   39.08%    passed     88.46%\n",
      "   7  251221.Q7NN83_GLOVI   39.96%    passed     94.06%\n",
      "   8  243231.Q74CP4_GEOSL   40.85%    passed     40.84%\n",
      "   9  83333.G3P1_ECOLI      41.21%    passed     85.54%\n",
      "  10  83333.E4PD_ECOLI      39.79%    passed     31.13%\n",
      "  11  289376.B5YI31_THEYD   40.67%    passed     86.36%\n",
      "  12  289376.B5YLA2_THEYD   27.53%    passed     44.61%\n",
      "  13  224324.G3P_AQUAE      39.25%    passed     85.78%\n",
      "  14  243273.G3P_MYCGE      40.14%    passed     60.55%\n",
      "  15  243090.Q7UVH8_RHOBA   39.25%    passed     99.40%\n",
      "  16  189518.Q8F5H7_LEPIN   40.50%    passed     72.59%\n",
      "  17  226186.Q89ZV8_BACTN   40.32%    passed     72.05%\n",
      "  18  1111708.G3P1_SYNY3    39.79%    passed     81.24%\n",
      "  19  1111708.G3P2_SYNY3    40.14%    passed     91.26%\n",
      "  20  224308.G3P2_BACSU     39.61%    passed     94.93%\n",
      "  21  224308.G3P1_BACSU     40.50%    passed     47.39%\n",
      "  22  83332.G3P_MYCTU       39.79%    passed     72.37%\n",
      "  23  100226.Q93J08_STRCO   41.03%    failed      0.16%\n",
      "  24  100226.Q9FC43_STRCO   14.56%    passed     56.48%\n",
      "  25  100226.G3P_STRCO      40.32%    passed     98.21%\n",
      "  26  324602.A9WAP0_CHLAA   40.50%    passed     36.50%\n",
      "  27  324602.A9WBV5_CHLAA   39.43%    passed     90.79%\n",
      "  28  208964.G3P_PSEAE      40.67%    passed     22.81%\n",
      "  29  208964.Q9HZK4_PSEAE   18.12%    passed     81.11%\n",
      "  30  208964.Q9I5Y5_PSEAE   37.30%    failed      0.10%\n",
      "  31  190304.Q8RFN9_FUSNN   40.50%    passed     73.03%\n",
      "  32  243230.Q9RUP1_DEIRA   41.39%    passed     95.98%\n",
      "  33  224911.Q89U94_BRADU   40.50%    passed     98.11%\n",
      "  34  515635.B8E2J2_DICTD   40.14%    passed     94.59%\n",
      "  35  243274.G3P_THEMA      40.85%    passed     67.75%\n",
      "  36  NP_213724.1           39.25%    passed     85.78%\n",
      "****  TOTAL                 38.49%  4 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "NOTE: NP_213724.1 is identical to 224324.G3P_AQUAE but kept for subsequent analysis\n",
      "\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.017 seconds\n",
      "NOTE: ModelFinder requires 27 MB RAM!\n",
      "ModelFinder will test 546 protein models (sample size: 563) ...\n",
      " No. Model         -LnL         df  AIC          AICc         BIC\n",
      "  1  Dayhoff       19233.441    69  38604.881    38624.476    38903.878\n",
      "  2  Dayhoff+I     18905.683    70  37951.366    37971.570    38254.696\n",
      "  3  Dayhoff+G4    18275.712    70  36691.424    36711.627    36994.753\n",
      "  4  Dayhoff+I+G4  18249.148    71  36640.295    36661.118    36947.958\n",
      "  5  Dayhoff+R2    18424.261    71  36990.522    37011.345    37298.185\n",
      "  6  Dayhoff+R3    18295.696    73  36737.391    36759.485    37053.721\n",
      "  7  Dayhoff+R4    18261.791    75  36673.583    36696.991    36998.579\n",
      "  8  Dayhoff+R5    18248.067    77  36650.135    36674.902    36983.797\n",
      "  9  Dayhoff+R6    18239.064    79  36636.127    36662.297    36978.457\n",
      " 10  Dayhoff+R7    18233.573    81  36629.145    36656.763    36980.141\n",
      " 14  Dayhoff+F     19292.756    88  38761.513    38794.559    39142.841\n",
      " 15  Dayhoff+F+I   18970.613    89  38119.226    38153.095    38504.888\n",
      " 16  Dayhoff+F+G4  18312.169    89  36802.338    36836.207    37188.000\n",
      " 17  Dayhoff+F+I+G4 18285.864    90  36751.727    36786.431    37141.723\n",
      " 18  Dayhoff+F+R2  18470.278    90  37120.556    37155.259    37510.551\n",
      " 19  Dayhoff+F+R3  18335.739    92  36855.478    36891.887    37254.140\n",
      " 20  Dayhoff+F+R4  18304.678    94  36797.357    36835.519    37204.685\n",
      " 21  Dayhoff+F+R5  18290.705    96  36773.410    36813.376    37189.405\n",
      " 22  Dayhoff+F+R6  18281.379    98  36758.758    36800.577    37183.419\n",
      " 23  Dayhoff+F+R7  18275.611    100 36751.222    36794.945    37184.550\n",
      " 27  mtMAM         21561.107    69  43260.214    43279.808    43559.210\n",
      " 28  mtMAM+I       21215.648    70  42571.295    42591.498    42874.625\n",
      " 29  mtMAM+G4      19475.349    70  39090.698    39110.902    39394.028\n",
      " 30  mtMAM+I+G4    19454.122    71  39050.243    39071.066    39357.906\n",
      " 31  mtMAM+R2      19870.275    71  39882.550    39903.372    40190.212\n",
      " 32  mtMAM+R3      19599.738    73  39345.476    39367.570    39661.805\n",
      " 33  mtMAM+R4      19546.634    75  39243.268    39266.677    39568.264\n",
      " 34  mtMAM+R5      19501.321    77  39156.643    39181.410    39490.305\n",
      " 35  mtMAM+R6      19475.158    79  39108.316    39134.486    39450.645\n",
      " 36  mtMAM+R7      19461.957    81  39085.914    39113.532    39436.910\n",
      " 37  mtMAM+R8      19451.482    83  39068.964    39098.075    39428.626\n",
      " 38  mtMAM+R9      19445.398    85  39060.795    39091.445    39429.124\n",
      " 40  mtMAM+F       20638.415    88  41452.830    41485.877    41834.159\n",
      " 41  mtMAM+F+I     20256.440    89  40690.879    40724.748    41076.541\n",
      " 42  mtMAM+F+G4    18892.185    89  37962.371    37996.240    38348.033\n",
      " 43  mtMAM+F+I+G4  18864.661    90  37909.323    37944.026    38299.318\n",
      " 44  mtMAM+F+R2    19243.744    90  38667.488    38702.192    39057.483\n",
      " 45  mtMAM+F+R3    19009.283    92  38202.566    38238.975    38601.228\n",
      " 46  mtMAM+F+R4    18960.516    94  38109.033    38147.195    38516.361\n",
      " 47  mtMAM+F+R5    18921.907    96  38035.814    38075.780    38451.809\n",
      " 48  mtMAM+F+R6    18901.408    98  37998.815    38040.634    38423.477\n",
      " 49  mtMAM+F+R7    18890.415    100 37980.830    38024.553    38414.158\n",
      " 50  mtMAM+F+R8    18880.125    102 37964.250    38009.929    38406.245\n",
      " 51  mtMAM+F+R9    18873.811    104 37955.621    38003.307    38406.282\n",
      " 53  JTT           19151.160    69  38440.321    38459.915    38739.317\n",
      " 54  JTT+I         18836.844    70  37813.687    37833.890    38117.017\n",
      " 55  JTT+G4        18192.342    70  36524.685    36544.888    36828.014\n",
      " 56  JTT+I+G4      18167.362    71  36476.724    36497.547    36784.387\n",
      " 57  JTT+R2        18347.418    71  36836.835    36857.658    37144.498\n",
      " 58  JTT+R3        18206.407    73  36558.814    36580.909    36875.144\n",
      " 59  JTT+R4        18180.972    75  36511.944    36535.353    36836.940\n",
      " 60  JTT+R5        18172.693    77  36499.386    36524.153    36833.049\n",
      " 61  JTT+R6        18164.873    79  36487.746    36513.916    36830.075\n",
      " 62  JTT+R7        18159.636    81  36481.272    36508.890    36832.268\n",
      " 66  JTT+F         19245.505    88  38667.009    38700.056    39048.338\n",
      " 67  JTT+F+I       18931.158    89  38040.316    38074.185    38425.978\n",
      " 68  JTT+F+G4      18248.646    89  36675.293    36709.162    37060.955\n",
      " 69  JTT+F+I+G4    18223.965    90  36627.930    36662.634    37017.925\n",
      " 70  JTT+F+R2      18413.804    90  37007.607    37042.311    37397.602\n",
      " 71  JTT+F+R3      18267.398    92  36718.796    36755.205    37117.458\n",
      " 72  JTT+F+R4      18241.368    94  36670.737    36708.899    37078.065\n",
      " 73  JTT+F+R5      18232.503    96  36657.007    36696.972    37073.001\n",
      " 74  JTT+F+R6      18224.097    98  36644.193    36686.012    37068.855\n",
      " 75  JTT+F+R7      18218.345    100 36636.690    36680.413    37070.018\n",
      " 79  WAG           18803.959    69  37745.917    37765.512    38044.914\n",
      " 80  WAG+I         18521.099    70  37182.198    37202.402    37485.528\n",
      " 81  WAG+G4        18020.482    70  36180.963    36201.166    36484.293\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 82  WAG+I+G4      17994.628    71  36131.256    36152.079    36438.919\n",
      " 83  WAG+R2        18129.585    71  36401.169    36421.992    36708.832\n",
      " 84  WAG+R3        18021.818    73  36189.636    36211.730    36505.965\n",
      " 85  WAG+R4        17990.853    75  36131.706    36155.114    36456.702\n",
      " 86  WAG+R5        17988.475    77  36130.951    36155.718    36464.613\n",
      " 92  WAG+F         18847.350    88  37870.700    37903.746    38252.028\n",
      " 93  WAG+F+I       18566.959    89  37311.918    37345.787    37697.580\n",
      " 94  WAG+F+G4      18048.400    89  36274.799    36308.668    36660.461\n",
      " 95  WAG+F+I+G4    18023.312    90  36226.625    36261.328    36616.620\n",
      " 96  WAG+F+R2      18166.484    90  36512.967    36547.671    36902.962\n",
      " 97  WAG+F+R3      18054.021    92  36292.041    36328.450    36690.703\n",
      " 98  WAG+F+R4      18017.539    94  36223.078    36261.241    36630.406\n",
      " 99  WAG+F+R5      18016.646    96  36225.292    36265.258    36641.287\n",
      "105  cpREV         18985.557    69  38109.113    38128.707    38408.109\n",
      "106  cpREV+I       18707.460    70  37554.921    37575.124    37858.250\n",
      "107  cpREV+G4      18152.336    70  36444.672    36464.875    36748.002\n",
      "108  cpREV+I+G4    18122.005    71  36386.010    36406.833    36693.673\n",
      "109  cpREV+R2      18279.685    71  36701.370    36722.193    37009.033\n",
      "110  cpREV+R3      18157.831    73  36461.661    36483.755    36777.991\n",
      "111  cpREV+R4      18123.833    75  36397.666    36421.074    36722.662\n",
      "112  cpREV+R5      18118.699    77  36391.399    36416.166    36725.061\n",
      "118  cpREV+F       19023.898    88  38223.797    38256.843    38605.125\n",
      "119  cpREV+F+I     18730.187    89  37638.375    37672.244    38024.037\n",
      "120  cpREV+F+G4    18149.419    89  36476.839    36510.708    36862.501\n",
      "121  cpREV+F+I+G4  18120.555    90  36421.110    36455.813    36811.105\n",
      "122  cpREV+F+R2    18284.273    90  36748.547    36783.250    37138.542\n",
      "123  cpREV+F+R3    18162.379    92  36508.757    36545.166    36907.419\n",
      "124  cpREV+F+R4    18118.603    94  36425.205    36463.367    36832.533\n",
      "125  cpREV+F+R5    18116.484    96  36424.968    36464.934    36840.963\n",
      "131  mtREV         20366.053    69  40870.107    40889.701    41169.103\n",
      "132  mtREV+I       20070.906    70  40281.812    40302.015    40585.141\n",
      "133  mtREV+G4      19036.051    70  38212.103    38232.306    38515.432\n",
      "134  mtREV+I+G4    19017.581    71  38177.162    38197.985    38484.825\n",
      "135  mtREV+R2      19286.923    71  38715.846    38736.669    39023.509\n",
      "136  mtREV+R3      19077.442    73  38300.884    38322.978    38617.213\n",
      "137  mtREV+R4      19026.897    75  38203.794    38227.202    38528.790\n",
      "138  mtREV+R5      19025.005    77  38204.010    38228.777    38537.673\n",
      "144  mtREV+F       19617.709    88  39411.417    39444.464    39792.746\n",
      "145  mtREV+F+I     19292.601    89  38763.202    38797.071    39148.864\n",
      "146  mtREV+F+G4    18424.163    89  37026.326    37060.195    37411.988\n",
      "147  mtREV+F+I+G4  18403.131    90  36986.262    37020.965    37376.257\n",
      "148  mtREV+F+R2    18656.495    90  37492.989    37527.693    37882.985\n",
      "149  mtREV+F+R3    18463.407    92  37110.814    37147.223    37509.476\n",
      "150  mtREV+F+R4    18412.710    94  37013.419    37051.582    37420.748\n",
      "151  mtREV+F+R5    18410.098    96  37012.196    37052.161    37428.190\n",
      "157  rtREV         19017.791    69  38173.581    38193.175    38472.577\n",
      "158  rtREV+I       18733.641    70  37607.282    37627.485    37910.611\n",
      "159  rtREV+G4      18105.670    70  36351.340    36371.543    36654.669\n",
      "160  rtREV+I+G4    18082.730    71  36307.460    36328.282    36615.122\n",
      "161  rtREV+R2      18261.689    71  36665.377    36686.200    36973.040\n",
      "162  rtREV+R3      18122.340    73  36390.679    36412.773    36707.009\n",
      "163  rtREV+R4      18081.710    75  36313.419    36336.828    36638.415\n",
      "164  rtREV+R5      18081.048    77  36316.096    36340.863    36649.759\n",
      "170  rtREV+F       18961.262    88  38098.525    38131.571    38479.853\n",
      "171  rtREV+F+I     18670.560    89  37519.120    37552.989    37904.782\n",
      "172  rtREV+F+G4    18006.341    89  36190.682    36224.551    36576.344\n",
      "173  rtREV+F+I+G4  17983.904    90  36147.807    36182.511    36537.802\n",
      "174  rtREV+F+R2    18170.092    90  36520.183    36554.887    36910.178\n",
      "175  rtREV+F+R3    18023.654    92  36231.307    36267.716    36629.969\n",
      "176  rtREV+F+R4    17984.370    94  36156.740    36194.903    36564.069\n",
      "177  rtREV+F+R5    17983.739    96  36159.478    36199.443    36575.473\n",
      "183  mtART         20953.485    69  42044.971    42064.565    42343.967\n",
      "184  mtART+I       20652.972    70  41445.945    41466.148    41749.274\n",
      "185  mtART+G4      19111.000    70  38362.001    38382.204    38665.331\n",
      "186  mtART+I+G4    19091.728    71  38325.456    38346.279    38633.119\n",
      "187  mtART+R2      19437.827    71  39017.653    39038.476    39325.316\n",
      "188  mtART+R3      19191.237    73  38528.473    38550.567    38844.803\n",
      "189  mtART+R4      19113.581    75  38377.163    38400.571    38702.159\n",
      "190  mtART+R5      19105.748    77  38365.497    38390.264    38699.159\n",
      "191  mtART+R6      19105.639    79  38369.278    38395.448    38711.607\n",
      "196  mtART+F       20110.626    88  40397.253    40430.299    40778.581\n",
      "197  mtART+F+I     19775.042    89  39728.084    39761.953    40113.746\n",
      "198  mtART+F+G4    18491.886    89  37161.772    37195.641    37547.434\n",
      "199  mtART+F+I+G4  18470.566    90  37121.131    37155.834    37511.126\n",
      "200  mtART+F+R2    18828.914    90  37837.828    37872.531    38227.823\n",
      "201  mtART+F+R3    18577.583    92  37339.167    37375.575    37737.828\n",
      "202  mtART+F+R4    18500.984    94  37189.968    37228.131    37597.296\n",
      "203  mtART+F+R5    18496.893    96  37185.786    37225.751    37601.781\n",
      "209  mtZOA         20331.524    69  40801.048    40820.643    41100.044\n",
      "210  mtZOA+I       20042.048    70  40224.097    40244.300    40527.426\n",
      "211  mtZOA+G4      18710.391    70  37560.782    37580.985    37864.111\n",
      "212  mtZOA+I+G4    18691.326    71  37524.652    37545.475    37832.315\n",
      "213  mtZOA+R2      18999.964    71  38141.929    38162.752    38449.592\n",
      "214  mtZOA+R3      18771.476    73  37688.951    37711.045    38005.281\n",
      "215  mtZOA+R4      18712.767    75  37575.534    37598.943    37900.530\n",
      "216  mtZOA+R5      18706.374    77  37566.749    37591.516    37900.411\n",
      "217  mtZOA+R6      18706.357    79  37570.714    37596.884    37913.043\n",
      "222  mtZOA+F       19768.896    88  39713.792    39746.839    40095.121\n",
      "223  mtZOA+F+I     19445.480    89  39068.959    39102.828    39454.621\n",
      "224  mtZOA+F+G4    18345.512    89  36869.024    36902.892    37254.685\n",
      "225  mtZOA+F+I+G4  18325.981    90  36831.962    36866.666    37221.958\n",
      "226  mtZOA+F+R2    18634.566    90  37449.132    37483.835    37839.127\n",
      "227  mtZOA+F+R3    18408.478    92  37000.955    37037.364    37399.617\n",
      "228  mtZOA+F+R4    18350.040    94  36888.079    36926.242    37295.408\n",
      "229  mtZOA+F+R5    18346.271    96  36884.542    36924.508    37300.537\n",
      "235  VT            18812.667    69  37763.334    37782.929    38062.331\n",
      "236  VT+I          18533.384    70  37206.767    37226.970    37510.097\n",
      "237  VT+G4         18037.519    70  36215.037    36235.241    36518.367\n",
      "238  VT+I+G4       18013.173    71  36168.346    36189.169    36476.009\n",
      "239  VT+R2         18149.347    71  36440.695    36461.518    36748.358\n",
      "240  VT+R3         18036.086    73  36218.173    36240.267    36534.502\n",
      "241  VT+R4         18007.050    75  36164.101    36187.509    36489.097\n",
      "242  VT+R5         18005.782    77  36165.564    36190.331    36499.226\n",
      "248  VT+F          18855.482    88  37886.964    37920.010    38268.292\n",
      "249  VT+F+I        18577.392    89  37332.783    37366.652    37718.445\n",
      "250  VT+F+G4       18064.523    89  36307.046    36340.914    36692.707\n",
      "251  VT+F+I+G4     18040.839    90  36261.679    36296.382    36651.674\n",
      "252  VT+F+R2       18182.964    90  36545.928    36580.631    36935.923\n",
      "253  VT+F+R3       18065.670    92  36315.339    36351.748    36714.001\n",
      "254  VT+F+R4       18035.433    94  36258.866    36297.028    36666.194\n",
      "255  VT+F+R5       18033.322    96  36258.643    36298.609    36674.638\n",
      "261  LG            18882.923    69  37903.846    37923.441    38202.843\n",
      "262  LG+I          18593.393    70  37326.786    37346.989    37630.115\n",
      "263  LG+G4         17948.908    70  36037.815    36058.019    36341.145\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "264  LG+I+G4       17927.679    71  35997.357    36018.180    36305.020\n",
      "265  LG+R2         18110.916    71  36363.831    36384.654    36671.494\n",
      "266  LG+R3         17964.514    73  36075.027    36097.121    36391.357\n",
      "267  LG+R4         17927.896    75  36005.791    36029.200    36330.787\n",
      "268  LG+R5         17924.500    77  36003.000    36027.767    36336.662\n",
      "274  LG+F          18953.923    88  38083.846    38116.893    38465.175\n",
      "275  LG+F+I        18662.904    89  37503.807    37537.676    37889.469\n",
      "276  LG+F+G4       17983.584    89  36145.167    36179.036    36530.829\n",
      "277  LG+F+I+G4     17963.037    90  36106.073    36140.776    36496.068\n",
      "278  LG+F+R2       18150.012    90  36480.023    36514.726    36870.018\n",
      "279  LG+F+R3       17999.707    92  36183.413    36219.822    36582.075\n",
      "280  LG+F+R4       17964.008    94  36116.016    36154.178    36523.344\n",
      "281  LG+F+R5       17960.695    96  36113.390    36153.356    36529.385\n",
      "287  DCMut         19230.280    69  38598.560    38618.155    38897.557\n",
      "288  DCMut+I       18903.005    70  37946.010    37966.213    38249.340\n",
      "289  DCMut+G4      18275.421    70  36690.841    36711.044    36994.171\n",
      "290  DCMut+I+G4    18248.769    71  36639.539    36660.362    36947.202\n",
      "291  DCMut+R2      18419.176    71  36980.351    37001.174    37288.014\n",
      "292  DCMut+R3      18292.380    73  36730.760    36752.855    37047.090\n",
      "293  DCMut+R4      18238.543    75  36627.087    36650.495    36952.083\n",
      "294  DCMut+R5      18235.707    77  36625.415    36650.182    36959.077\n",
      "300  DCMut+F       19288.901    88  38753.801    38786.848    39135.130\n",
      "301  DCMut+F+I     18967.237    89  38112.474    38146.343    38498.136\n",
      "302  DCMut+F+G4    18311.174    89  36800.348    36834.217    37186.010\n",
      "303  DCMut+F+I+G4  18284.849    90  36749.698    36784.401    37139.693\n",
      "304  DCMut+F+R2    18467.659    90  37115.319    37150.022    37505.314\n",
      "305  DCMut+F+R3    18332.534    92  36849.068    36885.476    37247.729\n",
      "306  DCMut+F+R4    18280.634    94  36749.267    36787.430    37156.596\n",
      "307  DCMut+F+R5    18277.554    96  36747.108    36787.074    37163.103\n",
      "313  PMB           18808.630    69  37755.260    37774.854    38054.256\n",
      "314  PMB+I         18552.846    70  37245.692    37265.896    37549.022\n",
      "315  PMB+G4        18086.159    70  36312.318    36332.521    36615.647\n",
      "316  PMB+I+G4      18062.239    71  36266.477    36287.300    36574.140\n",
      "317  PMB+R2        18194.108    71  36530.215    36551.038    36837.878\n",
      "318  PMB+R3        18087.100    73  36320.200    36342.294    36636.530\n",
      "319  PMB+R4        18056.070    75  36262.139    36285.548    36587.135\n",
      "320  PMB+R5        18055.561    77  36265.122    36289.889    36598.785\n",
      "326  PMB+F         18818.869    88  37813.738    37846.784    38195.067\n",
      "327  PMB+F+I       18557.147    89  37292.294    37326.163    37677.956\n",
      "328  PMB+F+G4      18079.782    89  36337.564    36371.433    36723.226\n",
      "329  PMB+F+I+G4    18055.438    90  36290.876    36325.580    36680.871\n",
      "330  PMB+F+R2      18191.940    90  36563.881    36598.584    36953.876\n",
      "331  PMB+F+R3      18082.494    92  36348.989    36385.397    36747.651\n",
      "332  PMB+F+R4      18049.804    94  36287.607    36325.769    36694.935\n",
      "333  PMB+F+R5      18048.889    96  36289.779    36329.744    36705.774\n",
      "339  HIVb          19853.620    69  39845.240    39864.835    40144.237\n",
      "340  HIVb+I        19474.940    70  39089.881    39110.084    39393.210\n",
      "341  HIVb+G4       18659.429    70  37458.858    37479.061    37762.188\n",
      "342  HIVb+I+G4     18630.757    71  37403.513    37424.336    37711.176\n",
      "343  HIVb+R2       18857.952    71  37857.903    37878.726    38165.566\n",
      "344  HIVb+R3       18686.224    73  37518.448    37540.542    37834.777\n",
      "345  HIVb+R4       18633.840    75  37417.680    37441.089    37742.676\n",
      "346  HIVb+R5       18627.950    77  37409.900    37434.667    37743.562\n",
      "352  HIVb+F        19927.929    88  40031.858    40064.905    40413.187\n",
      "353  HIVb+F+I      19535.723    89  39249.447    39283.316    39635.109\n",
      "354  HIVb+F+G4     18646.268    89  37470.536    37504.405    37856.198\n",
      "355  HIVb+F+I+G4   18619.352    90  37418.704    37453.408    37808.699\n",
      "356  HIVb+F+R2     18872.554    90  37925.109    37959.812    38315.104\n",
      "357  HIVb+F+R3     18678.844    92  37541.688    37578.097    37940.350\n",
      "358  HIVb+F+R4     18627.064    94  37442.128    37480.290    37849.456\n",
      "359  HIVb+F+R5     18621.663    96  37435.327    37475.292    37851.322\n",
      "365  HIVw          20555.719    69  41249.438    41269.032    41548.434\n",
      "366  HIVw+I        20199.206    70  40538.412    40558.615    40841.742\n",
      "367  HIVw+G4       19413.200    70  38966.400    38986.604    39269.730\n",
      "368  HIVw+I+G4     19380.784    71  38903.567    38924.390    39211.230\n",
      "369  HIVw+R2       19587.086    71  39316.173    39336.996    39623.836\n",
      "370  HIVw+R3       19436.385    73  39018.769    39040.863    39335.099\n",
      "371  HIVw+R4       19386.748    75  38923.495    38946.904    39248.491\n",
      "372  HIVw+R5       19381.359    77  38916.718    38941.485    39250.380\n",
      "378  HIVw+F        20373.692    88  40923.383    40956.430    41304.712\n",
      "379  HIVw+F+I      20002.179    89  40182.358    40216.227    40568.020\n",
      "380  HIVw+F+G4     19144.650    89  38467.300    38501.169    38852.962\n",
      "381  HIVw+F+I+G4   19115.031    90  38410.062    38444.766    38800.057\n",
      "382  HIVw+F+R2     19345.769    90  38871.539    38906.242    39261.534\n",
      "383  HIVw+F+R3     19181.129    92  38546.258    38582.666    38944.920\n",
      "384  HIVw+F+R4     19123.427    94  38434.854    38473.016    38842.182\n",
      "385  HIVw+F+R5     19118.313    96  38428.627    38468.593    38844.622\n",
      "391  JTTDCMut      19142.467    69  38422.933    38442.528    38721.930\n",
      "392  JTTDCMut+I    18828.502    70  37797.003    37817.207    38100.333\n",
      "393  JTTDCMut+G4   18188.527    70  36517.053    36537.257    36820.383\n",
      "394  JTTDCMut+I+G4 18163.259    71  36468.519    36489.341    36776.182\n",
      "395  JTTDCMut+R2   18343.162    71  36828.324    36849.147    37135.987\n",
      "396  JTTDCMut+R3   18201.579    73  36549.159    36571.253    36865.488\n",
      "397  JTTDCMut+R4   18160.939    75  36471.877    36495.286    36796.873\n",
      "398  JTTDCMut+R5   18158.314    77  36470.628    36495.395    36804.290\n",
      "404  JTTDCMut+F    19239.796    88  38655.593    38688.639    39036.921\n",
      "405  JTTDCMut+F+I  18925.670    89  38029.341    38063.210    38415.003\n",
      "406  JTTDCMut+F+G4 18245.907    89  36669.815    36703.684    37055.477\n",
      "407  JTTDCMut+F+I+G4 18221.042    90  36622.084    36656.787    37012.079\n",
      "408  JTTDCMut+F+R2 18411.469    90  37002.938    37037.641    37392.933\n",
      "409  JTTDCMut+F+R3 18264.546    92  36713.092    36749.501    37111.754\n",
      "410  JTTDCMut+F+R4 18222.264    94  36632.528    36670.690    37039.856\n",
      "411  JTTDCMut+F+R5 18219.781    96  36631.562    36671.528    37047.557\n",
      "417  FLU           19590.638    69  39319.275    39338.870    39618.272\n",
      "418  FLU+I         19278.563    70  38697.126    38717.330    39000.456\n",
      "419  FLU+G4        18530.102    70  37200.203    37220.406    37503.533\n",
      "420  FLU+I+G4      18505.678    71  37153.356    37174.179    37461.019\n",
      "421  FLU+R2        18696.536    71  37535.072    37555.895    37842.735\n",
      "422  FLU+R3        18552.734    73  37251.468    37273.562    37567.797\n",
      "423  FLU+R4        18509.469    75  37168.938    37192.346    37493.934\n",
      "424  FLU+R5        18508.095    77  37170.189    37194.956    37503.852\n",
      "430  FLU+F         19585.210    88  39346.420    39379.466    39727.749\n",
      "431  FLU+F+I       19267.169    89  38712.339    38746.208    39098.001\n",
      "432  FLU+F+G4      18469.741    89  37117.481    37151.350    37503.143\n",
      "433  FLU+F+I+G4    18442.418    90  37064.835    37099.538    37454.830\n",
      "434  FLU+F+R2      18647.173    90  37474.346    37509.049    37864.341\n",
      "435  FLU+F+R3      18493.372    92  37170.744    37207.152    37569.406\n",
      "436  FLU+F+R4      18454.883    94  37097.767    37135.929    37505.095\n",
      "437  FLU+F+R5      18452.934    96  37097.869    37137.835    37513.864\n",
      "443  Blosum62      18784.879    69  37707.757    37727.351    38006.753\n",
      "444  Blosum62+I    18535.410    70  37210.820    37231.023    37514.150\n",
      "445  Blosum62+G4   18048.663    70  36237.326    36257.529    36540.655\n",
      "446  Blosum62+I+G4 18026.229    71  36194.458    36215.281    36502.121\n",
      "447  Blosum62+R2   18165.886    71  36473.772    36494.595    36781.435\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "448  Blosum62+R3   18049.399    73  36244.798    36266.892    36561.127\n",
      "449  Blosum62+R4   18021.180    75  36192.360    36215.769    36517.356\n",
      "450  Blosum62+R5   18020.048    77  36194.096    36218.863    36527.758\n",
      "456  Blosum62+F    18811.880    88  37799.759    37832.806    38181.088\n",
      "457  Blosum62+F+I  18556.015    89  37290.029    37323.898    37675.691\n",
      "458  Blosum62+F+G4 18059.770    89  36297.541    36331.410    36683.202\n",
      "459  Blosum62+F+I+G4 18037.171    90  36254.341    36289.045    36644.336\n",
      "460  Blosum62+F+R2 18177.814    90  36535.629    36570.332    36925.624\n",
      "461  Blosum62+F+R3 18061.842    92  36307.684    36344.093    36706.346\n",
      "462  Blosum62+F+R4 18031.265    94  36250.529    36288.691    36657.857\n",
      "463  Blosum62+F+R5 18029.983    96  36251.965    36291.931    36667.960\n",
      "469  mtMet         20552.782    69  41243.563    41263.158    41542.560\n",
      "470  mtMet+I       20260.503    70  40661.006    40681.209    40964.335\n",
      "471  mtMet+G4      19089.035    70  38318.070    38338.273    38621.400\n",
      "472  mtMet+I+G4    19071.140    71  38284.281    38305.104    38591.944\n",
      "473  mtMet+R2      19356.522    71  38855.045    38875.868    39162.708\n",
      "474  mtMet+R3      19153.721    73  38453.442    38475.536    38769.772\n",
      "475  mtMet+R4      19088.749    75  38327.499    38350.907    38652.495\n",
      "476  mtMet+R5      19083.074    77  38320.147    38344.914    38653.810\n",
      "482  mtMet+F       19839.876    88  39855.752    39888.799    40237.081\n",
      "483  mtMet+F+I     19506.628    89  39191.256    39225.125    39576.918\n",
      "484  mtMet+F+G4    18477.680    89  37133.360    37167.229    37519.022\n",
      "485  mtMet+F+I+G4  18454.174    90  37088.349    37123.052    37478.344\n",
      "486  mtMet+F+R2    18750.534    90  37681.068    37715.771    38071.063\n",
      "487  mtMet+F+R3    18541.605    92  37267.210    37303.619    37665.872\n",
      "488  mtMet+F+R4    18478.728    94  37145.456    37183.619    37552.784\n",
      "489  mtMet+F+R5    18474.510    96  37141.020    37180.986    37557.015\n",
      "495  mtVer         20946.741    69  42031.482    42051.076    42330.478\n",
      "496  mtVer+I       20623.364    70  41386.728    41406.931    41690.058\n",
      "497  mtVer+G4      19251.169    70  38642.339    38662.542    38945.669\n",
      "498  mtVer+I+G4    19222.214    71  38586.427    38607.250    38894.090\n",
      "499  mtVer+R2      19530.950    71  39203.900    39224.722    39511.562\n",
      "500  mtVer+R3      19334.572    73  38815.145    38837.239    39131.474\n",
      "501  mtVer+R4      19250.642    75  38651.284    38674.693    38976.280\n",
      "502  mtVer+R5      19244.331    77  38642.662    38667.429    38976.325\n",
      "508  mtVer+F       20225.584    88  40627.169    40660.215    41008.497\n",
      "509  mtVer+F+I     19855.121    89  39888.242    39922.111    40273.904\n",
      "510  mtVer+F+G4    18720.372    89  37618.744    37652.613    38004.406\n",
      "511  mtVer+F+I+G4  18689.594    90  37559.187    37593.891    37949.182\n",
      "512  mtVer+F+R2    19012.930    90  38205.859    38240.563    38595.855\n",
      "513  mtVer+F+R3    18802.691    92  37789.382    37825.791    38188.044\n",
      "514  mtVer+F+R4    18728.671    94  37645.343    37683.505    38052.671\n",
      "515  mtVer+F+R5    18723.553    96  37639.106    37679.072    38055.101\n",
      "521  mtInv         20432.659    69  41003.319    41022.913    41302.315\n",
      "522  mtInv+I       20150.294    70  40440.588    40460.792    40743.918\n",
      "523  mtInv+G4      19141.660    70  38423.320    38443.523    38726.649\n",
      "524  mtInv+I+G4    19125.742    71  38393.483    38414.306    38701.146\n",
      "525  mtInv+R2      19382.168    71  38906.335    38927.158    39213.998\n",
      "526  mtInv+R3      19192.131    73  38530.262    38552.356    38846.592\n",
      "527  mtInv+R4      19137.437    75  38424.874    38448.282    38749.869\n",
      "528  mtInv+R5      19132.814    77  38419.628    38444.395    38753.290\n",
      "534  mtInv+F       19507.800    88  39191.600    39224.646    39572.929\n",
      "535  mtInv+F+I     19195.804    89  38569.609    38603.478    38955.271\n",
      "536  mtInv+F+G4    18308.590    89  36795.180    36829.049    37180.842\n",
      "537  mtInv+F+I+G4  18289.482    90  36758.964    36793.667    37148.959\n",
      "538  mtInv+F+R2    18549.954    90  37279.907    37314.610    37669.902\n",
      "539  mtInv+F+R3    18355.775    92  36895.551    36931.959    37294.212\n",
      "540  mtInv+F+R4    18305.995    94  36799.991    36838.153    37207.319\n",
      "541  mtInv+F+R5    18302.415    96  36796.830    36836.796    37212.825\n",
      "Akaike Information Criterion:           LG+I+G4\n",
      "Corrected Akaike Information Criterion: LG+I+G4\n",
      "Bayesian Information Criterion:         LG+I+G4\n",
      "Best-fit model: LG+I+G4 chosen according to BIC\n",
      "\n",
      "All model information printed to NP_213724.1.blast_homologs.alg.model.gz\n",
      "CPU time for ModelFinder: 731.162 seconds (0h:12m:11s)\n",
      "Wall-clock time for ModelFinder: 733.944 seconds (0h:12m:13s)\n",
      "\n",
      "NOTE: 10 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "Thoroughly optimizing +I+G parameters from 10 start values...\n",
      "Init pinv, alpha: 0.000, 1.000 / Estimate: 0.000, 0.858 / LogL: -17948.899\n",
      "Init pinv, alpha: 0.031, 1.000 / Estimate: 0.060, 1.259 / LogL: -17927.683\n",
      "Init pinv, alpha: 0.062, 1.000 / Estimate: 0.061, 1.261 / LogL: -17927.675\n",
      "Init pinv, alpha: 0.092, 1.000 / Estimate: 0.063, 1.263 / LogL: -17927.686\n",
      "Init pinv, alpha: 0.123, 1.000 / Estimate: 0.062, 1.262 / LogL: -17927.682\n",
      "Init pinv, alpha: 0.154, 1.000 / Estimate: 0.063, 1.262 / LogL: -17927.692\n",
      "Init pinv, alpha: 0.185, 1.000 / Estimate: 0.063, 1.263 / LogL: -17927.695\n",
      "Init pinv, alpha: 0.216, 1.000 / Estimate: 0.062, 1.261 / LogL: -17927.680\n",
      "Init pinv, alpha: 0.246, 1.000 / Estimate: 0.062, 1.261 / LogL: -17927.683\n",
      "Init pinv, alpha: 0.277, 1.000 / Estimate: 0.062, 1.261 / LogL: -17927.688\n",
      "Optimal pinv,alpha: 0.061, 1.261 / LogL: -17927.675\n",
      "\n",
      "Parameters optimization took 18.520 sec\n",
      "Computing ML distances based on estimated model parameters... 0.479 sec\n",
      "Computing BIONJ tree...\n",
      "0.001 seconds\n",
      "Log-likelihood of BIONJ tree: -17853.486\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 1.121 second\n",
      "Computing log-likelihood of 98 initial trees ... 18.008 seconds\n",
      "Current best score: -17853.486\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -17831.008\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 6: -17829.198\n",
      "Iteration 10 / LogL: -17852.111 / Time: 0h:0m:55s\n",
      "Iteration 20 / LogL: -17875.090 / Time: 0h:1m:11s\n",
      "Finish initializing candidate tree set (20)\n",
      "Current best tree score: -17829.198 / CPU time: 52.149\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 22: -17828.328\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 26: -17827.534\n",
      "Iteration 30 / LogL: -17856.481 / Time: 0h:1m:31s (0h:5m:3s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 33: -17824.178\n",
      "Iteration 40 / LogL: -17826.903 / Time: 0h:1m:47s (0h:4m:16s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 46: -17821.404\n",
      "Iteration 50 / LogL: -17824.292 / Time: 0h:2m:4s (0h:4m:3s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 55: -17821.339\n"
     ]
    }
   ],
   "source": [
    "# Build trees\n",
    "/home/miniconda3/bin/iqtree -s NP_213724.1.blast_homologs.alg\n",
    "/home/miniconda3/bin/iqtree -s NP_213887.1.blast_homologs.alg\n",
    "/home/miniconda3/bin/iqtree -s NP_Unk01.blast_homologs.alg\n",
    "/home/miniconda3/bin/iqtree -s NP_Unk02.blast_homologs.alg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br><br>\n",
    "### 3. Display trees\n",
    "Given that the phylogenetic trees will be visualized using ete3, a Python tool, I needed to create a separate Python script that can be executed as `xvfb-run python print_trees.py <treefile> additional_seq_info.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xvfb-run python print_trees.py NP_Unk01.blast_homologs.alg.treefile additional_seq_info.tsv\n",
    "xvfb-run python print_trees.py NP_Unk02.blast_homologs.alg.treefile additional_seq_info.tsv\n",
    "# running this commands raises an error and never shows the trees so I printed the trees running this program in my own python launcher"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Unk01 tree\n",
    "<img src=\"Unk01_tree.png\" alt=\"Drawing\" style=\"width: 500px;\"/>  \n",
    "#### Unk02 tree\n",
    "<img src=\"Unk02_tree.png\" alt=\"Drawing\" style=\"width: 500px;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br><br>\n",
    "### 4. Answers\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__What is the closest ortholog from a phylogenetic point of view? From what species?__ <br>\n",
    "\n",
    "- Unk01. The closest ortholog  is the 4Fe-4S iron sulfur cluster binding proteins from NifH/frxC family and belonging to the Arquea species Methanosarcina acetivorans <br>\n",
    "- NP_Unk02. The closest ortholog is norR, a PFAM sigma-54 factor interaction domain-containing protein from Aquifex aeolicus."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Do orthology assignment support your previous functional annotations?¬† (you might need to look up the functional annotation (i.e. gene names) of close orthologs)__ <br>\n",
    "\n",
    "- Unk01. This orthology assignment does support my previous functional annotation given that the protein‚Äôs function is also nitrogen binding <br>\n",
    "- NP_Unk02. According to my prediction, this protein was a œÉ-54 dependent transcriptional regulator, and its closest ortholog is protein domain that interacts with œÉ-54, so I think the prediction was accurate."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Are all genes present in the reference proteome of the same species? Why not? Do they all over expressed genes share the same evolutionary history?__ <br>\n",
    "\n",
    "No, in the reference proteome we have genes from many different species because in order to do a phylogenetic tree and infer orthologs you need to have sequences from different species to which you can compare you problem sequences. \n",
    "Not all the over expressed genes share the same evolutionary history. Unk01 seems to be closer to methanogenic Archaea while Unk02 is closer to other bacteria. It is posible that Unk01 is the result of horizontal gene transfer and that is why it seems to have such a different origin."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
